consensus/istanbul/backend/engine_test.go

```diff
@@ -506,3 +577,126 @@ func TestWriteCommittedSeals(t *testing.T) {
 		t.Errorf("error mismatch: have %v, want %v", err, errInvalidCommittedSeals)
 	}
 }
+
+func makeSnapshotTestConfigItems() []interface{} {
+	return []interface{}{
+		minimumStake(new(big.Int).SetUint64(5500000)),
+		stakingUpdateInterval(1),
+		proposerUpdateInterval(1),
+		proposerPolicy(params.WeightedRandom),
+	}
+}
+
+func makeFakeStakingInfo(blockNumber uint64, keys []*ecdsa.PrivateKey, amounts []uint64) *reward.StakingInfo {
+	stakingInfo := &reward.StakingInfo{
+		BlockNum: blockNumber,
+	}
+	for idx, key := range keys {
+		addr := crypto.PubkeyToAddress(key.PublicKey)
+
+		pk, _ := crypto.GenerateKey()
+		rewardAddr := crypto.PubkeyToAddress(pk.PublicKey)
+
+		stakingInfo.CouncilNodeAddrs = append(stakingInfo.CouncilNodeAddrs, addr)
+		stakingInfo.CouncilStakingAmounts = append(stakingInfo.CouncilStakingAmounts, amounts[idx])
+		stakingInfo.CouncilRewardAddrs = append(stakingInfo.CouncilRewardAddrs, rewardAddr)
+	}
+	return stakingInfo
+}
+
+func TestSnapshot(t *testing.T) {
+	type testcase struct {
+		stakingAmounts      []uint64 // test staking amounts of each validator
+		isIstanbulComatible bool     // whether or not if the inserted block is istanbul compatible
+		expectedValidators  int      // the number of expected validators
+		expectedDemoted     int      // the number of expected demoted validators
+	}
+
+	testcases := []testcase{
+		// The following testcases are the ones before istanbul incompatible change
+		{
+			[]uint64{5000000, 5000000, 5000000, 5000000},
```

Comment:
@jeongkyun-oh Could you add more edge cases like below?

The cases that the staking amount is ms-1, ms+1, ms, or 0.

If they are covered by other test cases, forget it. :)